instrumentation/opentelemetry-instrumentation-falcon/src/opentelemetry/instrumentation/falcon/__init__.py

```diff
@@ -461,8 +522,76 @@ class FalconInstrumentor(BaseInstrumentor):
     def instrumentation_dependencies(self) -> Collection[str]:
         return _instruments
 
-    # pylint:disable=no-self-use
-    def _remove_instrumented_middleware(self, app):
+    @staticmethod
+    def instrument_app(
+        app,
+        request_hook=None,
+        response_hook=None,
+        tracer_provider=None,
+        meter_provider=None,
+        excluded_urls=None,
+        **opts,
+    ):
+        if not hasattr(app, "_is_instrumented_by_opentelemetry"):
+
+            class FalconAPI(_InstrumentedFalconAPI):
+                def __init__(self, *args, **kwargs):
+                    for attribute in app.__slots__:
+                        setattr(self, attribute, getattr(app, attribute, None))
+                    self._otel_excluded_urls = (
+                        excluded_urls
+                        if excluded_urls is not None
+                        else get_excluded_urls("FALCON")
+                    )
+                    self._otel_tracer = trace.get_tracer(
+                        __name__, __version__, tracer_provider
+                    )
+                    self._otel_meter = get_meter(
+                        __name__, __version__, meter_provider
+                    )
+                    self.duration_histogram = self._otel_meter.create_histogram(
+                        name="http.server.duration",
+                        unit="ms",
+                        description="measures the duration of the inbound HTTP request",
+                    )
+                    self.active_requests_counter = self._otel_meter.create_up_down_counter(
+                        name="http.server.active_requests",
+                        unit="requests",
+                        description="measures the number of concurrent HTTP requests that are currently in-flight",
+                    )
+                    self._is_instrumented_by_opentelemetry = False
+
+            app = FalconAPI()
```

Comment:
I didn't understand this. The `app` arg passed is getting replaced by this; how doe this work?

Reply:
So, here we wanted to add more attributes in the original falcon api, but falcon do not let us add anything to already created app because of the use of [`__slots__`](https://towardsdatascience.com/understand-slots-in-python-e3081ef5196d#:~:text=However%2C%20__slots__%20will%20fix%20the%20attributes%20when%20you%20create%20the%20class.%20So%2C%20it%E2%80%99s%20not%20possible%20to%20add%20new%20attributes%20later.). So, what we did is, we created a new child class of Falcon API and copied all the attributes in the new objects with the additional attributes that we wanted to add and returned the new app instead of the original app.

Reply:
Does it contain everything? What happens to the rest of the attributes/properties/basically everything other than what `__slots__` have on `app` instance?

Reply:
Yes. Falcon can not have additional instance variables because of `__slots__`  and because the new object is of child class so all the methods of falcon class will be inherited.

Reply:
I am a little skeptical of this. I will try to find an example where this is not true.

Reply:
I think this is kind of sloppy. Try customising the base falcon and adding some instance variables, which will fail. For example, try this, and your manual instrumentation test fails with `AttributeError: 'FalconAPI' object has no attribute 'class_var'`

```patch
diff --git a/instrumentation/opentelemetry-instrumentation-falcon/tests/app.py b/instrumentation/opentelemetry-instrumentation-falcon/tests/app.py
index 3e4c62ec..ccac7113 100644
--- a/instrumentation/opentelemetry-instrumentation-falcon/tests/app.py
+++ b/instrumentation/opentelemetry-instrumentation-falcon/tests/app.py
@@ -64,11 +64,17 @@ class CustomResponseHeaderResource:
 def make_app():
     _parsed_falcon_version = package_version.parse(falcon.__version__)
     if _parsed_falcon_version < package_version.parse("3.0.0"):
+        class MyAPI(falcon.API):
+            class_var = "class_var"
+
         # Falcon 1 and Falcon 2
-        app = falcon.API()
+        app = MyAPI()
     else:
+        class MyApp(falcon.App):
+            class_var = "class_var"
+
         # Falcon 3
-        app = falcon.App()
+        app = MyApp()

     app.add_route("/hello", HelloWorldResource())
     app.add_route("/ping", HelloWorldResource())
diff --git a/instrumentation/opentelemetry-instrumentation-falcon/tests/test_falcon.py b/instrumentation/opentelemetry-instrumentation-falcon/tests/test_falcon.py
index df6f77c4..77eda353 100644
--- a/instrumentation/opentelemetry-instrumentation-falcon/tests/test_falcon.py
+++ b/instrumentation/opentelemetry-instrumentation-falcon/tests/test_falcon.py
@@ -272,6 +272,7 @@ class TestFalconInstrumentation(TestFalconBase, WsgiTestBase):
         self.client().simulate_get(path="/hello")
         spans = self.memory_exporter.get_finished_spans()
         self.assertEqual(len(spans), 0)
+        self.assertEqual(self.app.class_var, "class_var")

     def test_falcon_metrics(self):
         self.client().simulate_get("/hello/756")
@@ -394,6 +395,7 @@ class TestFalconManualInstrumentation(TestFalconInstrumentation):
         self.client().simulate_get(path="/hello")
         spans = self.memory_exporter.get_finished_spans()
         self.assertEqual(len(spans), 0)
+        self.assertEqual(self.app.class_var, "class_var")

     def test_metric_uninstrument(self):
         self.client().simulate_request(method="POST", path="/hello/756")
```

Reply:
Yes this is one possibility but this case is also not difficult to handle. I've added changes for that with the test you did. Please review.

Reply:
It's not about the difficulty of handling the case. We don't want to break people's application to make instrumentation work. The new try-except looks hacky to just address this comment.

Reply:
Yes, I understand your point. What are the other possibilities? Do you have any suggestion? I can not think of anything else.

Reply:
I would move the tracer/meter/instruments/config state to `_TraceMiddleware` since it's the one that's actually using them and then add it to middleware list.

Reply:
Not just `_TraceMiddleware` but `__call__` function from `_InstrumentedFalconAPI` also uses those states. If we move all these states to `_TraceMiddleware` then we'll need to move all the instrumentation logic to middleware. Are you suggesting that?

Reply:
I think `_InstrumentedFalconAPI` is a simple WSGI middleware. It's not necessary they maintain separate tracer/meter/instruments on it's own. It can use what will be used for the TraceMiddleware.

Reply:
Sorry I'm not able to understand. Can you please elaborate?